Add PostgreSQL and Oracle migration files for DB backed resource group manager

posulliv opened the pull request:
I was unable to add an automated test for Oracle migrations. I wanted to use the image used in other tests `wnameless/oracle-xe-11g-r2` but Flyway community edition will not run migrations with versions of Oracle that old. If someone knows of a newer Oracle image that can be used, I will be happy to add a test.

nineinchnick commented on plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V1__add_resource_groups_schema.sql:
```diff
@@ -0,0 +1,43 @@
+    hard_cpu_limit VARCHAR(128),
+    parent BIGINT,
+    environment VARCHAR(128),
+    FOREIGN KEY (parent) REFERENCES resource_groups (resource_group_id) ON DELETE CASCADE
```
Should there be indexes created for FK columns? It might be out of the scope of this PR though, as they are not present in the existing MySQL schema.

posulliv replied:
I can do this in a follow up PR for MySQL/PostgreSQL/Oracle if that is OK with you.

hashhar replied:
Resource group tables aren't expected to have enough entries to make the indexes useful.

nineinchnick reviewed (approved):
I found this image: https://hub.docker.com/r/gvenzl/oracle-xe. Looks decent and it's used by the Testcontainers project.

posulliv commented:
@nineinchnick thanks! I added a test using that Oracle image.

aalbu commented on plugin/trino-resource-group-managers/src/main/resources/db/migration/oracle/V1__add_resource_groups_schema.sql:
```diff
@@ -0,0 +1,44 @@
+    CHECK (name in ('cpu_quota_period'))
+);
+
+CREATE TABLE resource_groups (
```
Note that in Oracle, the database issues an implicit COMMIT statement before and after every DDL statement.  That means this migration is not transactional.  We should break it up into multiple files.

posulliv replied:
@aalbu thanks! I wasn't aware Oracle did not have transactional DDL. Updated to break the migration into multiple files.

hashhar replied:
Anything other than PostgreSQL doesn't have transactional DDL IIRC, please check if this change also needs to be done for MySQL.

posulliv replied:
@hashhar I already have it in place for MySQL. The only migration I have now that does all DDL in 1 migration is PostgreSQL.

aalbu commented on plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsFlywayMigration.java:
```diff
@@ -24,26 +26,24 @@
-                .withPassword("test");
-        mysqlServer.start();
-        jdbi = Jdbi.create(mysqlServer.getJdbcUrl(), mysqlServer.getUsername(), mysqlServer.getPassword());
+        startContainer();
```
Nit: Instead of a contract that requires this method to have a side effect (initializing `container`), it could have a return type of `JdbcDatabaseContainer<?>` and the assignment would be visible here, in the super class.

posulliv replied:
thanks @aalbu! I have applied this suggestion.

posulliv commented:
@aalbu thanks for the review! All suggestions have been applied.

hashhar commented on plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/FlywayMigration.java:
```diff
@@ -25,12 +25,26 @@ private FlywayMigration()
+        else if (configDbUrl.startsWith("jdbc:mysql")) {
+            return "/db/migration/mysql";
+        }
+        throw new IllegalArgumentException("Unexpected Url: " + configDbUrl);
```
Would be nice to mention that PostgreSQL, MySQL and Oracle are the only supported databases.

IMO the URL validation should be done in `DbResourceGroupConfig` via a `@PostConstruct public void validate() {...}` method (see existing examples in code).

hashhar commented on plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsFlywayMigration.java:
```diff
@@ -24,26 +26,24 @@
 
 @Test(singleThreaded = true)
-public class TestDbResourceGroupsFlywayMigration
+public abstract class TestDbResourceGroupsFlywayMigration
```
Rename to `BaseTestDbResource...`.

hashhar commented on plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsFlywayMigration.java:
```diff
@@ -24,26 +26,24 @@
 import static org.testng.Assert.assertEquals;
 
 @Test(singleThreaded = true)
```
Reminder to ensure that all subclasses include `Test(singleThreaded = true)` too since if the subclass doesn't have any `@Test` method then the single-threaded is lost. See `forceTestNgToRespectSingleThreaded` methods in existing code.

hashhar replied:
I see that it's added to Oracle tests in next commit.

hashhar commented on plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsMysqlFlywayMigration.java:
```diff
@@ -0,0 +1,87 @@
+    @Override
+    protected final JdbcDatabaseContainer<?> startContainer()
+    {
+        JdbcDatabaseContainer<?> container = new MySQLContainer<>("mysql:8.0.12");
```
@ebyhr @wendigo What do you think about starting to read image versions from a single properties file (or public constants) to ensure consistency in the versions being tested (easier to bump min and max tested versions that way).

@posulliv No change requested.

hashhar replied:
An alternative is to make versions injectable either from environment or system properties. That has the benefit that it's easier to test arbitrary versions via a matrix.

hashhar commented on plugin/trino-resource-group-managers/pom.xml:
```diff
@@ -119,6 +119,13 @@
+        <dependency>
+            <groupId>com.oracle.ojdbc</groupId>
+            <artifactId>ojdbc8</artifactId>
+            <version>${dep.oracle.version}</version>
```
Maybe we should extract this to root pom similar to MySQL and PostgreSQL?

hashhar replied:
On second thought, let's not do that just yet.

hashhar commented on plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsOracleFlywayMigration.java:
```diff
@@ -0,0 +1,93 @@
+public class TestDbResourceGroupsOracleFlywayMigration
+        extends TestDbResourceGroupsFlywayMigration
+{
+    @Test
+    public void forceTestNgToRespectSingleThreaded()
+    {
+        // TODO: Remove after updating TestNG to 7.4.0+ (https://github.com/trinodb/trino/issues/8571)
+        // TestNG doesn't enforce @Test(singleThreaded = true) when tests are defined in base class. According to
+        // https://github.com/cbeust/testng/issues/2361#issuecomment-688393166 a workaround it to add a dummy test to the leaf test class.
+    }
```
Move to bottom of class?

hashhar reviewed (commented):
Why does Postgres have a single migration? Is it because DDL is transactional in Postgres?

Looks good % comments.

Thanks for working on this @posulliv.

posulliv commented:
> Why does Postgres have a single migration? Is it because DDL is transactional in Postgres? (hint: should be a code comment or in the commit message - I'd prefer to have separate migrations just to make it easier to compare across databases but existing is fine as well).

Yes, that's correct. I agree though that it would be more consistent to have a separate migration per table. So I updated the PostgreSQL migrations to be 1 file per table.

hashhar reviewed (approved):
Thanks for adding this @posulliv .

hashhar commented:
Hmmm, seems like a test if failing now due to tests using H2.

```
Error:  io.trino.execution.resourcegroups.db.TestResourceGroupDbIntegration.testMemoryFraction  Time elapsed: 4.179 s  <<< FAILURE!
com.google.inject.CreationException: 
Unable to create injector, see the following errors:

1) Error notifying ProvisionListener LifeCycleModule$$Lambda$684/0x00000008004ee440 of DbResourceGroupConfig.
 Reason: LifeCycleStartException: Exception in PostConstruct method DbResourceGroupConfig::validate()
  at DbResourceGroupConfigurationManager.<init>(DbResourceGroupConfigurationManager.java:94)
      \_ for 2nd parameter config
  at H2ResourceGroupsModule.configure(H2ResourceGroupsModule.java:40)
  while locating DbResourceGroupConfigurationManager
  at H2ResourceGroupsModule.configure(H2ResourceGroupsModule.java:41)
  while locating ResourceGroupConfigurationManager
Caused by: LifeCycleStartException: Exception in PostConstruct method DbResourceGroupConfig::validate()
	at LifeCycleManager.startInstance(LifeCycleManager.java:246)
	at LifeCycleManager.addInstance(LifeCycleManager.java:212)
	at LifeCycleModule.provision(LifeCycleModule.java:62)
	at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
	at ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
	at ProviderInternalFactory.circularGet(ProviderInternalFactory.java:62)
	at InternalFactoryToInitializableAdapter.get(InternalFactoryToInitializableAdapter.java:47)
	at SingleParameterInjector.inject(SingleParameterInjector.java:40)
	at SingleParameterInjector.getAll(SingleParameterInjector.java:60)
	at ConstructorInjector.provision(ConstructorInjector.java:113)
	at ConstructorInjector.access$000(ConstructorInjector.java:33)
	at ConstructorInjector$1.call(ConstructorInjector.java:98)
	at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:109)
	at LifeCycleModule.provision(LifeCycleModule.java:54)
	at ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
	at ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
	at ConstructorInjector.construct(ConstructorInjector.java:93)
	at ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:296)
	at ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at SingletonScope$1.get(SingletonScope.java:169)
	at InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
	at FactoryProxy.get(FactoryProxy.java:60)
	at ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at SingletonScope$1.get(SingletonScope.java:169)
	at InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
	at InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:213)
	at InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
	at InternalInjectorCreator.build(InternalInjectorCreator.java:113)
	at Guice.createInjector(Guice.java:87)
	at Bootstrap.initialize(Bootstrap.java:269)
	at H2ResourceGroupConfigurationManagerFactory.create(H2ResourceGroupConfigurationManagerFactory.java:62)
	at InternalResourceGroupManager.setConfigurationManager(InternalResourceGroupManager.java:163)
	at H2TestUtil.createQueryRunner(H2TestUtil.java:137)
	at H2TestUtil.createQueryRunner(H2TestUtil.java:122)
	at H2TestUtil.getSimpleQueryRunner(H2TestUtil.java:154)
	at TestResourceGroupDbIntegration.testMemoryFraction(TestResourceGroupDbIntegration.java:28)
	at java.base/NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/Method.invoke(Method.java:566)
	at MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at Invoker.invokeMethod(Invoker.java:645)
	at Invoker.invokeTestMethod(Invoker.java:851)
	at Invoker.invokeTestMethods(Invoker.java:1177)
	at TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/Thread.run(Thread.java:829)
Caused by: IllegalArgumentException: Invalid JDBC URL: jdbc:h2:mem:test_1295574013934079806. Only PostgreSQL, MySQL, and Oracle are supported.
	at DbResourceGroupConfig.validate(DbResourceGroupConfig.java:104)
	at java.base/NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/Method.invoke(Method.java:566)
	at LifeCycleManager.startInstance(LifeCycleManager.java:241)
	... 48 more
```

cc: @posulliv can you fix this? I'm fine with reverting the change suggested in https://github.com/trinodb/trino/pull/9812#discussion_r785291381 and adding a comment why the validation is not done in the config class itself.

posulliv commented:
thanks @hashhar. I updated and added a comment. 

I forgot the unit tests use the h2 database. Perhaps some future work could be to update those unit tests to use one of the supported database engines instead of h2 and then we can add the validation back to the config class itself.

hashhar commented:
@aalbu / @nineinchnick Could you maybe take another look at the migrations being added since once added we won't get a chance to edit them - we can only add new migrations.

This looks good to me otherwise and ready to merge.

nineinchnick commented:
@hashhar LGTM

hashhar commented:
Thanks @posulliv for seeing this to the finish line.

Can you PTAL at suggested release note at https://github.com/trinodb/trino/issues/10552#issuecomment-1017359740? Feel free to suggest changes.

posulliv commented:
@hashhar there is an open PR from a long time back to add docs for this - https://github.com/trinodb/trino/pull/4523

Needs to be updated now. I will work on updating that docs PR.

Release note comment looks good to me. Thanks!!

mosabua commented:
Awesome @posulliv .. ping me when you want some input for the docs PR and anything related.
